File: DictionaryComressor/Application.cpp
#include "stdafx.h"
#include "IOHandler.h"
#include "Repository.h"
#include "Compressor.h"
#include "Application.h"
#include "CustomThread.h"

CAppManager::CAppManager(CIOHandler *manager)
	: m_iohandler(manager)
	, m_nextNumberThreadForWrite(0)
{
	m_repository = CRepository();
}

void CAppManager::SetInputFileName(std::string const &name)
{
	m_iohandler->SetInputFileName(name);
}

void CAppManager::SetOutputFileName(std::string const &name)
{
	m_iohandler->SetOutputFileName(name);
}

void CAppManager::ProcessFile()
{
	try
	{
		while (!m_iohandler->IsFileRead())
		{
			m_iohandler->PrepareEnvironment();

			for (size_t i = 0; i != CSingletonSystemInfo::GetInstance()->GetNumberOfProcessors(); ++i)
			{
				m_mthreads.push_back(new std::thread(&CAppManager::CompressFragment, this));
			}
			
			for (auto &thread : m_mthreads)
			{
				if ((*thread).joinable())
					thread->join();
			}

			m_mthreads.clear();
		}
		std::cout << "File processed" << std::endl;
		return;
	}
	catch (std::runtime_error const &ex)
	{
		std::cerr << "Error! " << ex.what() << std::endl;
		return;
	}
}

void CAppManager::SetNewTextFragment(CCompressor &compressor)
{
	m_repository.SetNewTextFragment(compressor.GetAllProcessingText());
}

void CAppManager::SaveResult()
{
	m_iohandler->OutputDictionary(m_repository.GetDictionary());
	m_iohandler->OutputProcessedText(m_repository.GetText());

	std::cout << "Results output in files" << std::endl;
}


void CAppManager::CompressFragment()
{
	bool success = true;

	while (success)
	{
		try
		{
			m_mutex.lock();
			CCompressor compressor(m_iohandler->GetOrder(), &m_repository, &m_compressorMutex);
			auto dataPtr = m_iohandler->GetViewMappingFile();
			m_mutex.unlock();

			compressor.SetTextFragment(dataPtr);
			compressor.SetLengthFragment(m_iohandler->GetSizeView());
			compressor.ProcessFragment();

			while (true)
			{
				if (compressor.GetOrder() == m_nextNumberThreadForWrite)
				{
					m_mutex.lock();
					SetNewTextFragment(compressor);
					m_mutex.unlock();
					++m_nextNumberThreadForWrite;
					break;
				}
				else
				{
					std::this_thread::sleep_for(std::chrono::seconds(1));
				}
			}
		}
		catch (std::runtime_error)
		{
			m_mutex.unlock();
			success = false;
		}
	}
}

CAppManager::~CAppManager()
{
	delete m_iohandler;
}

File: DictionaryCompressor/Application.cs
using System;
using System.Threading;
using System.Collections.Generic;

namespace DictionaryCompressor
{
	class Application
	{
		private IOManager m_iomanager;
		private RepositoryProcessedData m_myRepository;
		private int m_nextThreadOrderForWrite;
		private List<Thread> m_threads;
		private Mutex m_mutex;
		private Mutex m_mutexForCompressor;

		public Application(IOManager manager)
		{
			m_iomanager = manager;
			m_nextThreadOrderForWrite = 0;
			m_myRepository = new RepositoryProcessedData();
			m_threads = new List<Thread>();
			m_mutex = new Mutex();
			m_mutexForCompressor = new Mutex();
		}

		public void SetInputFileName(string name)
		{
			m_iomanager.SetInputFileName(name);
		}
		public void SetOutputFileName(string name)
		{
			m_iomanager.SetOutputFileName(name);
		}

		public void ProcessFile()
		{
			try
			{
				
				while (!m_iomanager.IsFileCompletelyReadOut())
				{
					m_iomanager.SettingInputFile();
					for (uint i = 0; i != Environment.ProcessorCount; ++i)
					{
						var thread = new Thread(ProcessNextFragment);
						thread.Start();

						m_threads.Add(thread);

					}

					foreach (var thread in m_threads)
					{
						if (thread.IsAlive)
						{
							thread.Join();
						}
					}

					m_threads.RemoveRange(0, m_threads.Count);
				}
				System.Console.WriteLine("File processed");
				return;
			}
			catch (Exception ex)
			{
			   System.Console.WriteLine("Error! " + ex.ToString());
				return;
			}
		}

		private void SaveNewInformation(Compressor compressor)
		{
			m_myRepository.SetNewTextFragment(compressor.GetAllProcessingText());
		}

		public void SaveResultsInFile()
		{
			m_iomanager.OutputDictionary(m_myRepository.GetDictionary());
			m_iomanager.OutputProcessedText(m_myRepository.GetText());
		}

		private void ProcessNextFragment()
		{
			var success = true;
			while (success)
			{
				try
				{
					m_mutex.WaitOne();
					Compressor compressor = new Compressor(m_iomanager.GetOrder(), m_myRepository, m_mutexForCompressor);
					var dataPtr = m_iomanager.GetViewMappingFile();
					m_mutex.ReleaseMutex();
					compressor.SetTextFragment(dataPtr);
					compressor.SetLengthFragment(m_iomanager.GetSizeView());
					compressor.EditFragment();

					while (true)
					{
						if (compressor.GetOrder() == m_nextThreadOrderForWrite)
						{
							m_mutex.WaitOne();
							SaveNewInformation(compressor);
							m_mutex.ReleaseMutex();
							++m_nextThreadOrderForWrite;
							break;
						}
						else
						{
							Thread.Sleep(1000);
						}
					}
				}
				catch (Exception)
				{
					m_mutex.ReleaseMutex();
					success = false;
				}
			}
		}
	}
}
